src/util/voxel.cpp

```diff
@@ -0,0 +1,48 @@
+    
+    
+    
+#include "voxel.hpp"
+#include "voxel_io.hpp"
+
+namespace voxel_utility {
+
+VoxelObject::VoxelObject(const std::filesystem::path path){
+    std::vector<uint32_t> voxel_colors;
+    try {
+        from_qb(path, voxel_colors, center_, size_);
+        ok_ = true;
+    } catch (const std::exception& e) {
+        LOG_ERROR(
+            logging::file_io_logger, "Could not create VoxelObject: {}", e.what()
+        );
+        ok_ = false;
+    }
+    // empty should always have index 0;
+    colors_.push_back(0);
+    for (std::size_t i = 0; i < voxel_colors.size(); i++){
+        uint16_t j = 0;
+        for (; j < colors_.size(); j++) {
+            if (colors_[j] == voxel_colors[i]){
+                goto on_break;
+            }
+        }
+        // no break
+        // voxel_colors[i] is not in colors
+        // we should add it to colors
+        // the index is the length of colors because it will be appended
+        j = colors_.size();
+        // add the color
+        colors_.push_back(voxel_colors[i]);
+        
+        on_break:
+        data_.push_back(j);
+
+    }
```

Comment:
Ok there are many things wrong here

First of all this is $O(n^2)$.

Second of all, a GOTO. REALLY.

```suggestion
    for (std::size_t i = 0; i < voxel_colors.size(); i++){
        size_t j = 0;
        for (; j < colors_.size(); j++) {
            if (colors_[j] == voxel_colors[i]){
                data_.push_back(j);
                break;
            }
        }
        // no break
        // voxel_colors[i] is not in colors
        // we should add it to colors
        // the index is the length of colors because it will be appended
        j = colors_.size();
        // add the color
        colors_.push_back(voxel_colors[i]);
        data_.push_back(j);
    }
```

Reply:
STAY MAD

Reply:
In all seriesness:
1) you broke the code. there are two for loops, and you only broke out of one.
2) I think n*n times is the best you will get. One could use a hash map and solve this in linear time, but that would require a hash map. I don't really want to implement that.